larray/core/metadata.py

```diff
@@ -67,23 +35,78 @@ class Metadata(BaseMetadata):
         >>> attrs = Metadata(title='the title')
 
         # add new metadata
-        >>> attrs['creation_date'] = datetime(2017, 2, 10)
+        >>> attrs.creation_date = datetime(2017, 2, 10)
 
         # access metadata
-        >>> attrs['creation_date']
+        >>> attrs.creation_date
         datetime.datetime(2017, 2, 10, 0, 0)
 
         # modify metadata
-        >>> attrs['creation_date'] = datetime(2017, 2, 16)
+        >>> attrs.creation_date = datetime(2017, 2, 16)
 
         # delete metadata
-        >>> del attrs['creation_date']
+        >>> del attrs.creation_date
         """
         def __init__(self, *args, **kwargs):
-            super(BaseMetadata, self).__init__(*args, **kwargs)
+            object.__setattr__(self, '__odict', OrderedDict(*args, **kwargs))
+
+        def __getattr__(self, key):
+            od = object.__getattribute__(self, '__odict')
+            if hasattr(od, key):
+                return getattr(od, key)
+            else:
+                try:
+                    return od[key]
+                except KeyError:
+                    raise AttributeError(key)
+
+        def __setattr__(self, key, value):
+            od = object.__getattribute__(self, '__odict')
+            od[key] = value
+
+        def __delattr__(self, key):
+            od = object.__getattribute__(self, '__odict')
+            del od[key]
+
+        def __dir__(self):
+            od = object.__getattribute__(self, '__odict')
+            return list(set(dir(self.__class__)) | set(self.__dict__.keys()) | set(od.keys()))
+
+        def method_factory(name):
+            fullname = '__%s__' % name
+            odict_method = getattr(OrderedDict, fullname)
+            def method(self, *args, **kwargs):
+                od = object.__getattribute__(self, '__odict')
+                return odict_method(od, *args, **kwargs)
+            return method
+
+        __getitem__ = method_factory('getitem')
+        __setitem__ = method_factory('setitem')
+        __delitem__ = method_factory('delitem')
+        __contains__ = method_factory('contains')
+
+        __eq__ = method_factory('eq')
+        __ne__ = method_factory('ne')
+        __ge__ = method_factory('ge')
+        __gt__ = method_factory('gt')
+        __le__ = method_factory('le')
+        __lt__ = method_factory('lt')
+
+        __iter__ = method_factory('iter')
+        __len__ = method_factory('len')
+
+        __reduce__ = method_factory('reduce')
+        __reduce_ex__ = method_factory('reduce_ex')
+
+        __reversed__ = method_factory('reversed')
+
+        __sizeof__ = method_factory('sizeof')
```

Comment:
Did you test sizeof, reduce and reduce_ex? They are probably broken (sizeof probably underevaluates the size and reduce might not work). If they don't work completely, it's better to remove them.